src/Utils/obiDialogParser.ts

```diff
@@ -112,51 +123,141 @@ export class ObiDialogParser {
         return luMap
     }
 
-    private async getTrainDialogsfromOBIDialog(obiDialog: OBITypes.OBIDialog): Promise<CLM.TrainDialog[]> {
-
-        let trainDialogs: CLM.TrainDialog[] = []
+    /**
+     * Walks the dialog tree from the given node.  Validates types of nodes in the tree and returns an
+     * in-memory representation of the tree.
+     * Tree construction is slightly complicated since child nodes can be referenced in `rules` or `steps`.
+     */
+    private async collectDialogNodes(obiDialog: OBITypes.OBIDialog): Promise<ObiDialogNode> {
+        let node: ObiDialogNode = new ObiDialogNode(obiDialog)
         if (obiDialog.rules) {
-            for (const rule of obiDialog.rules) {
-                if (rule.$type === OBIRuleType.INTENT_RULE) {
+            await this.collectDialogRuleChildren(node, obiDialog.rules)
+        }
+        if (obiDialog.steps) {
+            await this.collectDialogStepChildren(node, obiDialog.steps)
+        }
+        return node
+    }
 
-                    const textVariations = this.getTextVariations(rule.intent!)
-                    const extractorStep: CLM.TrainExtractorStep = {
-                        textVariations: textVariations
-                    }
-                    if (!rule.steps) {
+    // Collects dialog tree nodes from `Microsoft.IntentRule` elements in the dialog `rules` section.
+    private async collectDialogRuleChildren(node: ObiDialogNode, rules: OBITypes.MicrosoftIRule[]) {
+        for (const rule of rules) {
+            if (rule.$type !== OBIRuleType.INTENT_RULE) {
+                console.log(`Unhandled OBI rule type: ${rule.$type}`)
+                continue
+            }
+            const intent = rule.intent
+            if (!intent) {
+                throw new Error(`Rule is missing intent property`)
+            }
+            if (!rule.steps) {
+                continue
+            }
+            for (const step of rule.steps) {
+                if (typeof step === "string") {
+                    throw new Error("Unexpected string step")
+                }
+                if (step.$type !== OBIStepType.BEGIN_DIALOG || typeof step.dialog !== "string") {
+                    console.log(`Unhandled OBI Type: ${step.$type}`)
```

Comment:
Generally think about whether any of these console logs or throws should be warnings